### PR TITLE
[clang-format] Add cmake target clang-format-style-options for updating ClangFormatStyleOptions.rst

### PR DESCRIPTION
**Pull request by ichaer**
* Create a new `clang-format-style-options` build target which re-generates ClangFormatStyleOptions.rst from its source header files.

As discussed in https://github.com/llvm/llvm-project/pull/96804#discussion_r1718407404

### CODE REVIEW

**Comment by ichaer** [comment]
Sample failure: https://github.com/llvm/llvm-project/actions/runs/11233429933/job/31227131778?pr=111513#step:10:127

**Comment by ichaer** [comment]
@owenca, see what you think.

**Review by boomanaiden154 — CHANGES_REQUESTED** [review]
The documentation build workflow is not really the correct place to put this.

Also, why can't we just run the python script as part of building the docs instead of committing the result of the script to the repository?

**Comment by ichaer** [comment]
> The documentation build workflow is not really the correct place to put this.

Yeah, I agree, but I couldn't find a place which looked correct :sweat_smile:. This thing of adding built artifacts to the repo always introduces sadness... the documentation build workflow looked like the least bad place. I'll gladly take input on this!

> Also, why can't we just run the python script as part of building the docs instead of committing the result of the script to the repository?

That's an excellent question, maybe @owenca will know! I should have asked in https://github.com/llvm/llvm-project/pull/96804#discussion_r1718407404; what I did instead was try to reason out an explanation based on what I could see, and I came up with two theories. The first one is that we want for `ClangFormatStyleOptions.rst` to be available to anyone browsing the repo... it's not very convincing, but it doesn't sound absurd. I could definitely get behind my second theory, though: we want to force everyone to review `ClangFormatStyleOptions.rst` before merging changes to the header files from which it is produced, and putting it in the repo is a good way to inject that into the code review process.

Anyhow, I'm proposing this because in https://github.com/llvm/llvm-project/pull/96804#discussion_r1718407404 I thought that it was an unreasonable burden for project maintainers having to manually gate on changes which may affect `ClangFormatStyleOptions.rst`, and also an unnecessary source of friction during code reviews, so I volunteered to automate that part.

**Comment by boomanaiden154** [comment]
> That's an excellent question, maybe @owenca will know! I should have asked in https://github.com/llvm/llvm-project/pull/96804#discussion_r1718407404; what I did instead was try to reason out an explanation based on what I could see, and I came up with two theories. The first one is that we want for ClangFormatStyleOptions.rst to be available to anyone browsing the repo... it's not very convincing, but it doesn't sound absurd. I could definitely get behind my second theory, though: we want to force everyone to review ClangFormatStyleOptions.rst before merging changes to the header files from which it is produced, and putting it in the repo is a good way to inject that into the code review process.

Based on the commit history (d278e0eee3f8c8f25131983979062b3d8e94ae77, https://reviews.llvm.org/D1597), it seems like the real answer might be that it just never got implemented. That looks like it was from the time where there were two build systems that both needed to be maintained at the same time, so maybe somewhat understandable it wasn't implemented that way. There is precedent for generating docs from other files (several clang docs pages are generated through tablegen. So I'd prefer to see a patch doing that rather than just asserting that it is the same. I don't think there should be issues with the Python requirement either. Most importantly, we already require a relatively recent Python in CMake,  so its guaranteed to be available, and secondly building `clang-format` as part of a bootstrapping process doesn't make much sense.

That is my read on it, although I'm not a `clang-format` maintainer, so they might have more context there that I am not aware of.

> Anyhow, I'm proposing this because in https://github.com/llvm/llvm-project/pull/96804#discussion_r1718407404 I thought that it was an unreasonable burden for project maintainers having to manually gate on changes which may affect ClangFormatStyleOptions.rst, and also an unnecessary source of friction during code reviews, so I volunteered to automate that part.

Yeah, I saw the thread. Thanks for looking into process improvements here.

**Comment by ichaer** [comment]
> So I'd prefer to see a patch doing that rather than just asserting that it is the same.

@boomanaiden154, I'm willing to do that, but could we confirm with a clang-format maintainer first? In the course of that code review I referenced [I did introduce an error in the documentation through a valid change to the header files](https://github.com/llvm/llvm-project/pull/96804#discussion_r1707474893), and @mydeveloperday was able to catch it by looking at the file generated. The error I introduced would have been hard to catch through automation. I suppose the same could be said for all of the human-consumable artifacts, all the docs, but my impression is that this is a special case, user-facing documentation produced from comments in the source code.

My instinct is to agree with you and not version-control artifacts, I just want to make sure that the folks who are dealing with that part of the codebase more frequently agree that change would be net positive.

**Comment by owenca** [comment]
> Also, why can't we just run the python script as part of building the docs instead of committing the result of the script to the repository?

Because only part of the doc is generated by `dump_format_style.py`. See https://github.com/llvm/llvm-project/blob/1809d0fa1c15b16ca94381d8be3ef70c4a83c36b/clang/docs/ClangFormatStyleOptions.rst?plain=1#L3.

**Comment by owenca** [comment]
> @owenca, see what you think.

IMO it would be more helpful to add the build target to `lib/Format/CMakeLists.txt` instead. For example,
```
$ ninja FormatTests
[1/1] cd /Users/Owen/llvm-project/clang....12/bin/python3.12 dump_format_style.p
$ 
```

**Comment by ichaer** [comment]
> IMO it would be more helpful to add the build target to lib/Format/CMakeLists.txt instead. For example,
```
$ ninja FormatTests
```

Moving the target from `clang/docs/CMakeLists.txt` to `lib/Format/CMakeLists.txt` sounds great to me, but the testing part of the thing relies on us being in a `git clone`, it's a very CI-oriented thing, my impression is that it's more appropriate for it to live in the workflow yaml files. I'm saying this because you suggested `FormatTests` as the name... I'm not sure if `clang-format-style-options` is a good name, but I think it's closer to what we can do there. What do you think?

> I'm ok with not adding the CI check though.

IMO that would be sad, I would really like to free maintainers up from having to remember this dependency themselves. Is this because you feel `.github/workflows/docs.yml` is the wrong place for that validation? I see that step more as a pre-build validation for docs than as a test, that's why `docs.yml` looks like a good home for it from my point of view, but we could move it elsewhere if you're uncomfortable with that.

**Comment by owenca** [comment]
> > IMO it would be more helpful to add the build target to lib/Format/CMakeLists.txt instead. For example,
> 
> ```
> $ ninja FormatTests
> ```
> 
> Moving the target from `clang/docs/CMakeLists.txt` to `lib/Format/CMakeLists.txt` sounds great to me, but the testing part of the thing relies on us being in a `git clone`, it's a very CI-oriented thing, my impression is that it's more appropriate for it to live in the workflow yaml files. I'm saying this because you suggested `FormatTests` as the name... I'm not sure if `clang-format-style-options` is a good name, but I think it's closer to what we can do there. What do you think?

I wasn't suggesting a different name as `FormatTests` is an existing build target. See https://github.com/llvm/llvm-project/blob/0fba8381d2a71ff440fdf0ae30d59a0bf07fea75/clang/unittests/Format/CMakeLists.txt#L41

Actually, I like the `clang-format-style-options` name.

Anyway, my current workflow is:
1. Edit files in `clang/lib/Format` and `clang/unittests/Format`.
2. Edit `clang/include/clang/Format/Format.h` if needed.
3. Run `ninja FormatTests`.
4. Run `cd clang/docs/tools && dump_format_style.py` if Step 2 above wasn't skipped.
5. Run the unit tests.
6. Run `git commit -a`.

It would be nice if I didn't have to run step 4 manually.

> > I'm ok with not adding the CI check though.
> 
> IMO that would be sad, I would really like to free maintainers up from having to remember this dependency themselves. Is this because you feel `.github/workflows/docs.yml` is the wrong place for that validation? I see that step more as a pre-build validation for docs than as a test, that's why `docs.yml` looks like a good home for it from my point of view, but we could move it elsewhere if you're uncomfortable with that.

I'm not opposed to adding the CI check either.

**Comment by ichaer** [comment]
> It would be nice if I didn't have to run step 4 manually.

Got it! So ideally we'd move `clang-format-style-options` lower down the stack, to a target shared between `clang-format` and `FormatTests`.

I've made it a dependency of the `clangFormat` library, let me know what you think :).

**Comment by boomanaiden154** [comment]
> Because only part of the doc is generated by dump_format_style.py.

It should still be possible to only include the non-generated parts in the repo and then modify the script to add all the generated components on afterwards as a build step? Not sure if I'm missing something here.

**Comment by owenca** [comment]
> > Because only part of the doc is generated by dump_format_style.py.
> 
> It should still be possible to only include the non-generated parts in the repo and then modify the script to add all the generated components on afterwards as a build step? Not sure if I'm missing something here.

Sometimes we need to edit the non-generated parts, and it would be easier to have the entire contexts. Anyway, don't we need the generated RST file in the repo so that the corresponding HTML file (e.g. https://clang.llvm.org/docs/ClangFormatStyleOptions.html) will be readily available?

File: clang/lib/Format/CMakeLists.txt
Comment (owenca) [inline clang/lib/Format/CMakeLists.txt:61]: Please use `CLANG_SOURCE_DIR`. (See 78e7d9592d7c3b500ae2bae097d8cd7cd9369671.)

File: clang/lib/Format/CMakeLists.txt
Comment (owenca) [inline clang/lib/Format/CMakeLists.txt:58]: It seems that the command always runs even if the header files are not outdated.

**Review by owenca — COMMENTED** [review]
@ichaer I suggest that you split the workflow part to another patch so that we can get the cmake part approved.

**Comment by boomanaiden154** [comment]
#113739 should fix this problem by just adding the options as part of the build and omitting them from the documentation source.

File: clang/lib/Format/CMakeLists.txt
Comment (ichaer) [reply clang/lib/Format/CMakeLists.txt]: Sorry for the silence, @owenca, crazy days!

Your suggestion is much cleaner, thank you :)

**Comment by ichaer** [comment]
> @ichaer I suggest that you split the workflow part to another patch so that we can get the cmake part approved.

It's a bit sad, but I get your point, let's make it more focused.

File: clang/lib/Format/CMakeLists.txt
Comment (owenca) [inline clang/lib/Format/CMakeLists.txt:60]:
```suggestion
  WORKING_DIRECTORY ${docs_tools_dir}
```

**Review by owenca — APPROVED** [review]

File: clang/lib/Format/CMakeLists.txt
Comment (owenca) [reply clang/lib/Format/CMakeLists.txt:58]: Resolved in 6bec1806c9cc90f6e72fc04698f4221c86c5f95e.

File: clang/lib/Format/CMakeLists.txt
Comment (owenca) [reply clang/lib/Format/CMakeLists.txt:60]: Ditto.

File: clang/lib/Format/CMakeLists.txt
Comment (owenca) [reply clang/lib/Format/CMakeLists.txt:63]: Ditto.

**Comment by nikic** [comment]
Reverted this because it breaks a the build. I initially assumed this is due to disabled docs per the previous comment, but I think in my case the problem is actually that this seems to be trying to write to the source directory, which is forbidden. If you need to write out files, that has to happen in the build directory.
```
FAILED: /var/llvm-compile-time-tracker/llvm-project/clang/docs/ClangFormatStyleOptions.rst 
cd /var/llvm-compile-time-tracker/llvm-project/clang/docs/tools && /usr/bin/python3.10 dump_format_style.py
Traceback (most recent call last):
  File "/var/llvm-compile-time-tracker/llvm-project/clang/docs/tools/dump_format_style.py", line 23, in <module>
    with open(PLURALS_FILE, "a+") as f:
PermissionError: [Errno 13] Permission denied: '/var/llvm-compile-time-tracker/llvm-project/clang/docs/tools/plurals.txt'
```

**Comment by nikic** [comment]
Reading back here, the general thing you are trying to do here is not allowed. At best you can add this as a manually invoked target that is not implicitly called by anything else. Though the actually correct way to handle this is what was proposed in https://github.com/llvm/llvm-project/pull/113739.

**Comment by owenca** [comment]
> ```
> FAILED: /var/llvm-compile-time-tracker/llvm-project/clang/docs/ClangFormatStyleOptions.rst 
> cd /var/llvm-compile-time-tracker/llvm-project/clang/docs/tools && /usr/bin/python3.10 dump_format_style.py
> Traceback (most recent call last):
>   File "/var/llvm-compile-time-tracker/llvm-project/clang/docs/tools/dump_format_style.py", line 23, in <module>
>     with open(PLURALS_FILE, "a+") as f:
> PermissionError: [Errno 13] Permission denied: '/var/llvm-compile-time-tracker/llvm-project/clang/docs/tools/plurals.txt'
> ```

Changed the mode to read-only in 74d29c6393df606e98e7c42b9a97f56f335e3ffb.

**Comment by owenca** [comment]
> If CLANG_INCLUDE_DOCS=OFF then this should be skipped, otherwise the build will fail if llvm/clang/docs doesn't exist.

This was fixed in f0b09dfd4c84fe8ffbc726da39ee5255e1a22a2f, which also addressed an issue that ClangFormatStyleOptions.rst would be deleted by `ninja clean`.

**Pull request merged**
